sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs

```diff
@@ -282,57 +282,89 @@ private static void CloseLink(RequestResponseAmqpLink link)
             {
                 if (!_receiveLink.TryGetOpenedObject(out link))
                 {
-                    link = await _receiveLink.GetOrCreateAsync(UseMinimum(_connectionScope.SessionTimeout, timeout)).ConfigureAwait(false);
+                    link = await _receiveLink.GetOrCreateAsync(UseMinimum(_connectionScope.SessionTimeout, timeout))
+                        .ConfigureAwait(false);
                 }
+
                 cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
 
-                var messagesReceived = await Task.Factory
-                    .FromAsync<(ReceivingAmqpLink, int, TimeSpan?, TimeSpan), IEnumerable<AmqpMessage>>
-                (
-                    static (arguments, callback, state) =>
-                    {
-                        var (link, maxMessages, maxWaitTime, timeout) = arguments;
-                        return link.BeginReceiveRemoteMessages(
-                            maxMessages,
-                            TimeSpan.FromMilliseconds(20),
-                            maxWaitTime ?? timeout,
-                            callback,
-                            link);
-                    },
-                    static asyncResult =>
-                    {
-                        var link = (ReceivingAmqpLink)asyncResult.AsyncState;
-                        bool received = link.EndReceiveMessages(asyncResult, out IEnumerable<AmqpMessage> amqpMessages);
-                        return received ? amqpMessages : Enumerable.Empty<AmqpMessage>();
-                    },
-                    (link, maxMessages, maxWaitTime, timeout),
-                    default
+                var receiveMessagesCompletionSource =
```

Comment:
Unfortunately, I think we may need to revert this. @jsquire pointed out that with this approach we will just be leaving receive operations hanging on the AMQP link, which will cause a backup. I confirmed this with a test that attempts to receive after a previous cancel. I think we will need to either limit the scope of this change to just StopProcessing calls, because in that case it is okay that receive operations are blocked, or better yet, see if we can contribute Cancellation token support to the AMQP library. 
/cc @xinchen10

Reply:
```cs
        public async Task CancellingDoesNotBlockSubsequentReceives(bool prefetch)
        {
            await using (var scope = await ServiceBusScope.CreateWithQueue(enablePartitioning: false, enableSession: false))
            {
                await using var client = CreateClient();

                ServiceBusSender sender = client.CreateSender(scope.QueueName);
                var receiver = client.CreateReceiver(scope.QueueName, new ServiceBusReceiverOptions { PrefetchCount = prefetch ? 10 : 0 });

                using var cancellationTokenSource = new CancellationTokenSource(2000);
                var start = DateTime.UtcNow;

                Assert.That(
                    async () => await receiver.ReceiveMessageAsync(TimeSpan.FromSeconds(60), cancellationToken: cancellationTokenSource.Token),
                    Throws.InstanceOf<TaskCanceledException>());

                await sender.SendMessageAsync(GetMessage());
                var msg = await receiver.ReceiveMessageAsync();
                Assert.AreEqual(1, msg.DeliveryCount);
                var end = DateTime.UtcNow;
                Assert.NotNull(msg);
                Assert.Less(end - start, TimeSpan.FromSeconds(5));
            }
        }
```

Reply:
The above test fails on the second receive call as we are blocked on the cancelled receive.

Reply:
That's why we originally closed the link in the other PR but that has also other drawbacks. I think even StopProcessing can be problematic because the processor is designed to be restarted right?

Reply:
Yes, the processor can be restarted - actually StopProcessing, just stops receiving rather than closing any links. Close/Dispose would close links. I really think the best way forward is to try to get this integrated into the AMQP lib, so that we can actually end the operations early instead of ignoring them.

Reply:
I have tried to have this conversation already a year ago and failed. 

Given that we are stateful in the receiver we could discover that we previously canceled. Force close the link and reestablish in case of further receive attempts on the same link?

Reply:
The problem is that with sessionful receivers, attempting to re-establish the link may not work as it may be locked by another receiver.

Reply:
Cancellation tokens typically do not work well with network communication calls. From the AMQP perspective, I haven't found a real use case where someone needs to randomly cancel receive calls. It makes perfect sense to cancel it in the shutdown or close process, and that can be achieved by just closing the link. I agree that sometimes one may want to keep the link open but stop reading messages from it. In fact it is a perfect case supported by AMQP with the link credits and optionally the drain capability. With the current the AMQP library, one can stop the link by setting the credit to 0 but not be able to cancel the pending receive calls. I am open to add this support but not necessarily through cancellation tokens.

There are some options to achieve this.
1. `AmqpLink.SetTotalCredit`: we can enhance this API by cancelling pending receive calls when `totalCredit` is 0. 
2. Add a `Stop/Start` method to ReceivingAmqpLink to make it more explicit.

BTW for session receiver, to cancel a pending accept session call, the SB SDK could just close the link, even though it is still in the opening state. The OpenAsync call should return with OperationCancelledException.

Reply:
I would vote for #2.

Reply:
For 1, I don't think cancelling pending receive calls when totalCredit is 0 is sufficient because there could be concurrent receives occurring on the same link.

Reply:
I really have a hard time to understand all the push back against cancellationtoken. Cooperative cancellation is the defacto standard in dotnet for IO bound operations. It is present almost anywhere in moderns async enabled API even in the runtime as well as across the ecosystem. Even the SDK guidance of the whole azure SDK where a lot of people have contributed to and intense user studies have been done adheres to those principles because this is how this ecosystem works. So why so much push back?

Reply:
@danielmarbach the push back is not for cancellation tokens. Its more about supporting the shutdown scenario in a better way that also makes sense to AMQP (I admit that I am influenced more by other AMQP implementations, especially the Apache Qpid products and their JMS implementation). Your PR to the AMQP library (thank you for that) adds cancellation token to the receive method only. It gives a feeling that the library API is created on a needed basis and it was done just to make the shutdown scenario work. To properly support cancellation tokens, we will also need to look at other Task based APIs.

Reply:
Fair enough. Unfortunately, I'm only a community contributor without corporate backing so the only thing I could commit to in my precious spare time was exactly that. Your comment put things under a different light, and it sounds more like the door is open rather than the door is closed which I have potentially unrightfully experienced or shall I say read into the conversations. I appreciate you taking the time and clarify that.

If there would be some way to openly share on the repo some plans, ideas, directions including things that could be done I'm happy to contribute a few things when I have time, and it fits my small knowledge area that I have of the AMQP lib. For me, it boils down to have this project under some sort of active governance and communication plan to see where things are heading to (or not).